front: rename some trainId utils

clarani opened the pull request:
See commits :pray:

clarani committed:
front: rename isPacedTrain -> isPacedTrainId

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: rename isOccurrence -> isOccurrenceId

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: rename isTrainSchedule -> isTrainScheduleId

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: rename formatEditoastTrainIdToTrainScheduleId -> formatEditoastIdToTrainScheduleId

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: rename formatEditoastTrainIdToPacedTrainId -> formatEditoastIdToPacedTrainId

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: rename formatEditoastTrainIdToIndexedOccurrenceId and formatEditoastTrainIdToExceptionId utils

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: rename formatTrainScheduleIdToEditoastTrainId -> extractEditoastIdFromTrainScheduleId

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: rename formatPacedTrainIdToEditoastTrainId -> extractEditoastIdFromPacedTrainId

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: rename getOccurrenceIndexFromOccurrenceId -> extractOccurrenceIndexFromOccurrenceId

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: rename getExceptionIdFromOccurrenceId -> extractExceptionIdFromOccurrenceId

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

emersion commented:
Renaming the various `isXXXId()` functions is a no-brainer. Looks good!

For the conversion functions, I had a somewhat different proposal in mind, although your proposal is a good alternative. I'd like to call "raw" the IDs on the wire (instead of "editoast"), to save a few keystrokes, not tie the name to the software component which happens to provide the REST API and make it clear that these shouldn't be manipulated directly apart from API requests. Often times, the wording used to convert from/to raw values is "encode" (to raw) and "decode" (from raw). Putting all of this together:

- `formatEditoastTrainIdToTrainScheduleId()` → `decodeTrainScheduleId()`
- `formatTrainScheduleIdToEditoastTrainId()` → `encodeTrainScheduleId()`
- etc

After all, converting a `number` to a `TrainScheduleId` is not very different from converting a `string` to a `Date` object. Some other common alternative to encode/decode is "parse"/"format", although it's less symmetrical and the term "format" is overloaded/overused in our codebase (kudos for using a different word in your proposal!).

A perfectly valid criticism is that this proposal's wording is too terse and it's too easy to mix up the two sides. I'd reply that remembering what's "on the wire" and "we don't want to work with raw values" should help, and typing helps too.

Some trickier cases:

- `getExceptionIdFromOccurrenceId()`: I'd drop the "get" (→ `exceptionIdFromOccurrenceId()`), but I know some developers prefer to always keep a verb in function names. It's tricky to come up with a reasonable shorter name here.
- `formatOccurrenceIdToEditoastTrainId()`: your proposal renames this to `extractEditoastIdFromOccurrenceId()`, but I find this a bit confusing. The returned ID doesn't reference an occurrence, it references a paced train (a parent object). I'd suggest removing it, and providing a `pacedTrainIdFromOccurrenceId()` which returns a `PacedTrainId`. It's slightly more function calls when making an API request, but "encode"/"decode"/"raw" would be kept strictly for API requests, while `pacedTrainIdFromOccurrenceId()` would be more generally useful (and type-safe since it doesn't deal with raw `number` values).
- "simplify formatEditoastIdToIndexedOccurrenceId's signature": I prefer the old signature. Since both arguments are `number`s, it's very easy to mix them up and tricky to catch in code reviews. "simplify formatEditoastIdToExceptionId's signature" is less of an issue, but I would still keep the old signature for consistency.

clarani commented:
Thanks for your feedback :pray: 

We discussed this PR during the Front meeting, and here are the key points:
- We agreed that the terms `encode/decode` were not very explicit and could be confusing. Even though `formatEditoastIdToTrainScheduleId` and `extractEditoastIdFromTrainScheduleId` are longer, they're more descriptive and easier to understand for newcomers. We decided it's better to stick with my initial proposal.
- We also agreed that keeping verbs like `extract` or `format` (even if it makes the function names longer) improves clarity and is worth the trade-off.
- I’ve updated the function `formatOccurrenceIdToEditoastTrainId` to return a `pacedTrainId` instead of an editoast id, as you requested :+1:
- Regarding the simplification of the signatures for `formatEditoastIdToIndexedOccurrenceId` and `formatEditoastIdToExceptionId`, we concluded that the original versions were more explicit and less error-prone, so I've dropped the last two commits.

I'm going to merge this PR like this, but we can still continue to discuss it next week :)

clarani committed:
front: remove formatOccurrenceIdToEditoastTrainId and extractPacedTrainIdFromOccurenceId

extractPacedTrainIdFromOccurenceId was a duplicated of extractPacedTrainIdFromOccurrenceId.

The function formatOccurrenceIdToEditoastTrainId was not very useful and could be replaced by
the combination of extractPacedTrainIdFromOccurrenceId and extractEditoastIdFromPacedTrain.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

SharglutDev commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -19,10 +19,10 @@ import type {
-  formatTrainScheduleIdToEditoastTrainId,
-  isPacedTrain,
+  formatEditoastIdToPacedTrainId,
+  extractEditoastIdFromPacedTrainId,
+  extractEditoastIdFromTrainScheduleId,
```
super nitty : we could order the imports alphabetically :)

clarani replied:
Done :)

SharglutDev commented on front/src/utils/trainId.ts:
```diff
@@ -171,16 +151,26 @@ export const extractPacedTrainIdFromOccurenceId = (occurrenceId: OccurrenceId):
  * extract its paced train id with a PacedTrainId format (used across the front).
  */
 export const extractPacedTrainIdFromOccurrenceId = (occurrenceId: OccurrenceId): PacedTrainId => {
```
nit : technically, now that the id format changed, we are extracting the editoast id not the paced train id. PacedTrainId is then formatted.
I would have voted for `formatPacedTrainIdToOccurrenceId`. But I can live with that if you prefer to keep this naming :)

clarani replied:
I prefer to keep extract because we are still extracting a data from the input (and losing a part of it too). Thanks for the comment, I resolve to merge :pray:

SharglutDev reviewed (approved):
Lgtm and tested ✅ Just two nit comments, you can close them when fixed / not fixed

(Just need to update `isOccurrence` in `OccurrenceItem.tsx`)
